domain/application/state/charm.go

```diff
@@ -40,37 +40,41 @@ func NewCharmState(factory database.TxnRunnerFactory) *CharmState {
 	}
 }
 
-// GetCharmIDByRevision returns the charm ID by the natural key, for a
-// specific revision.
+// GetCharmID returns the charm ID by the natural key, for a
+// specific revision and source.
 // If the charm does not exist, a [errors.CharmNotFound] error is returned.
-func (s *CharmState) GetCharmIDByRevision(ctx context.Context, name string, revision int) (corecharm.ID, error) {
+func (s *CharmState) GetCharmID(ctx context.Context, name string, revision int, source charm.CharmSource) (corecharm.ID, error) {
 	db, err := s.DB()
 	if err != nil {
 		return "", internalerrors.Capture(err)
 	}
 
 	var ident charmID
-	args := charmReferenceNameRevision{
+	charmRef := charmReferenceNameRevision{
 		ReferenceName: name,
 		Revision:      revision,
+		Source:        string(source),
 	}
 
 	query := `
 SELECT &charmID.*
 FROM charm
-INNER JOIN charm_origin
-ON charm.uuid = charm_origin.charm_uuid
-WHERE charm_origin.reference_name = $charmReferenceNameRevision.reference_name
-AND charm_origin.revision = $charmReferenceNameRevision.revision;
+INNER JOIN charm_origin AS co
+ON charm.uuid = co.charm_uuid
+INNER JOIN charm_source AS cs
+ON cs.id = co.source_id
+WHERE co.reference_name = $charmReferenceNameRevision.reference_name
+AND co.revision = $charmReferenceNameRevision.revision
+AND cs.name = $charmReferenceNameRevision.source;
```

Comment:
This 100% should be a view.

Reply:
Yeah you're absolutely right, I changed things a little bit so we use the (already available) `v_charm_origin` view which has all 3 needed fields. Thanks for the catch!